### DISCUSSION TITLE
Upgrading from 3.0.2 to 3.1.5 causes draws to be at 0.0.

### DISCUSSION THREAD

**Discussion by ChadJessup** [SixLabors/ImageSharp#2816] [Q&A] [answered]
I'm having a hard time hunting down where the issue is, it could be my code somewhere, but I've spent a good chunk of time trying to hunt this issue down.

If you look at the first image, this is multiple assets draw on top of others - this is for 3.0.2.
![image](https://github.com/user-attachments/assets/777ce66f-75c1-4d4d-8393-9379b1d9bdfe)

After a 3.1.5 upgrade and no other changes, various components are now getting written at 0.0.
![image](https://github.com/user-attachments/assets/334ac90a-74b9-4745-899d-cc705e3e5deb)

My guess is that the minor version bump might have introduced new DrawImage overrides, and I'm now calling different code on accident since I didn't always use named parameters when calling into ImageSharp.

I should note that this first showed up when upgrading from 3.0.2 to 3.1.0 - but due to how input validation changed I had to make code changes to compile, but with 3.1.3 I don't have to do any code changes at all from 3.0.2.

```csharp
   public void BlitSurfaceToSurface(
       Image<Rgba32> src,
       SurfaceType dst,
       Point dstPoint,
       VO_BLT bltFlags = VO_BLT.SRCTRANSPARENCY,
       bool debug = false)
   {
       var dstSurface = this.Surfaces.SurfaceByTypes[dst];

       Rectangle dstRectangle = new()
       {
           Height = src.Height,
           Width = src.Width,
           X = dstPoint.X,
           Y = dstPoint.Y,
       };

       dstSurface.Image.Mutate(ctx =>
       {
           ctx.DrawImage(
               src,
               dstPoint,
               dstRectangle,
               colorBlending: PixelColorBlendingMode.Normal,
               //                alphaComposition: PixelAlphaCompositionMode.Dest,
               opacity: 1.0f);
       });
   }
```

On 3.1.5 Visual Studio is strongly implying I'm calling into this DrawImage override:

```csharp
    /// <summary>
    /// Draws the given image together with the currently processing image by blending their pixels.
    /// </summary>
    /// <param name="source">The current image processing context.</param>
    /// <param name="foreground">The image to draw on the currently processing image.</param>
    /// <param name="backgroundLocation">The location on the currently processing image at which to draw.</param>
    /// <param name="foregroundRectangle">The rectangle structure that specifies the portion of the image to draw.</param>
    /// <param name="colorBlending">The color blending to apply.</param>
    /// <param name="opacity">The opacity of the image to draw. Must be between 0 and 1.</param>
    /// <returns>The <see cref="IImageProcessingContext"/>.</returns>
    public static IImageProcessingContext DrawImage(
        this IImageProcessingContext source,
        Image foreground,
        Point backgroundLocation,
        Rectangle foregroundRectangle,
        PixelColorBlendingMode colorBlending,
        float opacity)
        => DrawImage(source, foreground, backgroundLocation, foregroundRectangle, colorBlending, source.GetGraphicsOptions().AlphaCompositionMode, opacity);
```

In 3.0.2 it looks like I'm calling the same thing (below) - I know in 3.1.0, I had to explicitly try to hit specific overrides because of the validation changes, but I'm trying to minimize changes for this discussion.

```csharp

    /// <summary>
    /// Draws the given image together with the currently processing image by blending their pixels.
    /// </summary>
    /// <param name="source">The current image processing context.</param>
    /// <param name="foreground">The image to draw on the currently processing image.</param>
    /// <param name="backgroundLocation">The location on the currently processing image at which to draw.</param>
    /// <param name="foregroundRectangle">The rectangle structure that specifies the portion of the image to draw.</param>
    /// <param name="colorBlending">The color blending to apply.</param>
    /// <param name="opacity">The opacity of the image to draw. Must be between 0 and 1.</param>
    /// <returns>The <see cref="IImageProcessingContext"/>.</returns>
    public static IImageProcessingContext DrawImage(
        this IImageProcessingContext source,
        Image foreground,
        Point backgroundLocation,
        Rectangle foregroundRectangle,
        PixelColorBlendingMode colorBlending,
        float opacity)
        => DrawImage(source, foreground, backgroundLocation, foregroundRectangle, colorBlending, source.GetGraphicsOptions().AlphaCompositionMode, opacity);
```

Digging through the releases, I have a gut feeling I'm being caught up in one of the DrawImage changes?

https://github.com/SixLabors/ImageSharp/pull/2474
https://github.com/SixLabors/ImageSharp/pull/2673

Any help would be appreciated, this seems like a pretty common code path to be hit, so I'm fairly certain it's an issue on my end, but not sure what's going on.

Thanks!

**Comment by ChadJessup** [comment]
For completeness, 3.1.0 gives this exception in the same line of code, but that exception is gone after 3.1.3...

![image](https://github.com/user-attachments/assets/bde564ec-b20e-454f-ad91-4d67997e4c4d)

**Reply by JimBobSquarePants** [reply, marked as answer]
Sorry @ChadJessup I really meant to look at this sooner. 

There was a lot of confusion plus some broken behavior prior to [v3.1.1](https://github.com/SixLabors/ImageSharp/releases/tag/v3.1.1) which were causing issues. I believe we have solved them all though now.

In your code example you are using the `dstPoint` parameter to populate both the location on the background at which to draw the foreground and the location on the foreground rectangle at which to sample the foreground.

```c#
Rectangle dstRectangle = new()
{
    Height = src.Height,
    Width = src.Width,
    X = dstPoint.X,
    Y = dstPoint.Y,
};

dstSurface.Image.Mutate(ctx =>
{
    ctx.DrawImage(
        src,
        dstPoint,
        dstRectangle,
        colorBlending: PixelColorBlendingMode.Normal,
        //                alphaComposition: PixelAlphaCompositionMode.Dest,
        opacity: 1.0f);
});
```

I can demonstrate using the following image and code how the API is designed to work. 

![issue_2447](https://github.com/user-attachments/assets/4f5f0d0f-d3be-43c0-a251-23e55a383f9d)

```c#
using Image<Rgba32> foreground = Image.Load<Rgba32>(@"D:\SixLabors\ImageSharp\tests\Images\Input\Png\issues\issue_2447.png");
using Image<Rgba32> background = new Image<Rgba32>(100, 100, Color.HotPink);

//                                                 // Draw here             // Draw this
background.Mutate(c => c.DrawImage(foreground, new Point(30, 30), new Rectangle(32, 32, 32, 32), 1F));

background.Save(@"C:\Users\james\Downloads\2186-draw-image.png");

```

![2186-draw-image](https://github.com/user-attachments/assets/d8305e80-79ee-4ca8-942c-9e99e478cf1c)

Does this make sense?

**Reply by ChadJessup** [reply]
Thanks for your help!

After swapping and adjusting some values based on your help, things seem to be lining up as expected now.

Sounds like I got it to work on an older version of ImageSharp, and worked with the bugs in the early 3.x versions, but that caused me to be broken when updating to new versions with the bug fixes. So, I coded against bugs, not the ideal way. :)

**Reply by JimBobSquarePants** [reply]
You wouldn't have to code against them if I didn't write them! 😁

I think it's all solid now so there shouldn't be any future surprises.
